digits/model/tasks/caffe_train.py

```diff
@@ -203,36 +212,58 @@ def save_files_classification(self):
             train_data_layer.top.append('data')
             train_data_layer.top.append('label')
             train_data_layer.include.add(phase = caffe_pb2.TRAIN)
-            train_data_layer.data_param.batch_size = constants.DEFAULT_BATCH_SIZE
+            if dataset_backend == 'lmdb':
+                train_data_layer.data_param.batch_size = constants.DEFAULT_BATCH_SIZE
+            elif dataset_backend == 'hdf5':
+                train_data_layer.hdf5_data_param.batch_size = constants.DEFAULT_BATCH_SIZE
             if self.crop_size:
+                assert dataset_backend != 'hdf5', 'HDF5Data layer does not support cropping'
                 train_data_layer.transform_param.crop_size = self.crop_size
             if has_val_set:
-                val_data_layer = train_val_network.layer.add(type = 'Data', name = 'data')
+                val_data_layer = train_val_network.layer.add(type = layer_type, name = 'data')
                 val_data_layer.top.append('data')
                 val_data_layer.top.append('label')
                 val_data_layer.include.add(phase = caffe_pb2.TEST)
-                val_data_layer.data_param.batch_size = constants.DEFAULT_BATCH_SIZE
+                if dataset_backend == 'lmdb':
+                    val_data_layer.data_param.batch_size = constants.DEFAULT_BATCH_SIZE
+                elif dataset_backend == 'hdf5':
+                    val_data_layer.hdf5_data_param.batch_size = constants.DEFAULT_BATCH_SIZE
                 if self.crop_size:
                     val_data_layer.transform_param.crop_size = self.crop_size
-        train_data_layer.data_param.source = self.dataset.path(self.dataset.train_db_task().db_name)
-        train_data_layer.data_param.backend = caffe_pb2.DataParameter.LMDB
-        if val_data_layer is not None and has_val_set:
-            val_data_layer.data_param.source = self.dataset.path(self.dataset.val_db_task().db_name)
-            val_data_layer.data_param.backend = caffe_pb2.DataParameter.LMDB
+        if dataset_backend == 'lmdb':
+            train_data_layer.data_param.source = self.dataset.path(self.dataset.train_db_task().db_name)
+            train_data_layer.data_param.backend = caffe_pb2.DataParameter.LMDB
+            if val_data_layer is not None and has_val_set:
+                val_data_layer.data_param.source = self.dataset.path(self.dataset.val_db_task().db_name)
+                val_data_layer.data_param.backend = caffe_pb2.DataParameter.LMDB
+        elif dataset_backend == 'hdf5':
+            train_data_layer.hdf5_data_param.source = self.dataset.path(self.dataset.train_db_task().textfile)
+            if val_data_layer is not None and has_val_set:
+                val_data_layer.hdf5_data_param.source = self.dataset.path(self.dataset.val_db_task().textfile)
         if self.use_mean:
+            assert dataset_backend != 'hdf5', 'HDF5Data layer does not support mean subtraction'
```

Comment:
Somebody needs to either add this functionality to the HDF5 layer or create a new "transformation" layer that still runs in the "prefetch" phase. See https://github.com/BVLC/caffe/issues/2840.

My current solution is to just throw an error from the DIGITS side which is more legible than a Caffe error.